src/messaging/ReliableMessageManager.cpp

```diff
@@ -0,0 +1,656 @@
+            ChipLogProgress(ExchangeManager, "ReliableMessageManager::ExpireTicks set nextRetransTimeTick to %u",
+                            RetransTable[i].nextRetransTimeTick);
+#endif
+        } // rc entry is allocated
+    }
+
+    // Re-Adjust the base time stamp to the most recent tick boundary
+    mTimeStampBase += (deltaTicks << mTimerIntervalShift);
+
+#if defined(RMP_TICKLESS_DEBUG)
+    ChipLogProgress(ExchangeManager, "ReliableMessageManager::ExpireTicks mTimeStampBase to %" PRIu64, mTimeStampBase);
+#endif
+}
+
+/**
+ * Handle physical wakeup of system due to ReliableMessageProtocol wakeup.
+ *
+ */
+void ReliableMessageManager::Timeout(System::Layer * aSystemLayer, void * aAppState, System::Error aError)
+{
+    ReliableMessageManager * manager = reinterpret_cast<ReliableMessageManager *>(aAppState);
+
+    VerifyOrDie((aSystemLayer != nullptr) && (manager != nullptr));
+
+#if defined(RMP_TICKLESS_DEBUG)
+    ChipLogProgress(ExchangeManager, "ReliableMessageManager::Timeout\n");
+#endif
+
+    // Make sure all tick counts are sync'd to the current time
+    manager->ExpireTicks();
+
+    // Execute any actions that are due this tick
+    manager->ExecuteActions();
+
+    // Calculate next physical wakeup
+    manager->StartTimer();
+}
+
+/**
+ *  Add a CHIP message into the retransmission table to be subsequently resent if a corresponding acknowledgment
+ *  is not received within the retransmission timeout.
+ *
+ *  @param[in]    rc        A pointer to the ExchangeContext object.
+ *
+ *  @param[in]    msgBuf    A pointer to the message buffer holding the CHIP message to be retransmitted.
+ *
+ *  @param[in]    messageId The message identifier of the stored CHIP message.
+ *
+ *  @param[out]   rEntry    A pointer to a pointer of a retransmission table entry added into the table.
+ *
+ *  @retval  #CHIP_ERROR_RETRANS_TABLE_FULL If there is no empty slot left in the table for addition.
+ *  @retval  #CHIP_NO_ERROR On success.
+ *
+ */
+CHIP_ERROR ReliableMessageManager::AddToRetransTable(ReliableMessageContext * rc, System::PacketBuffer * msgBuf, uint32_t messageId,
+                                                     uint16_t msgSendFlags, RetransTableEntry ** rEntry)
+{
+    bool added     = false;
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    for (int i = 0; i < CHIP_CONFIG_RMP_RETRANS_TABLE_SIZE; i++)
+    {
+        // Check the exchContext pointer for finding an empty slot in Table
+        if (!RetransTable[i].rc)
+        {
+            // Expire any virtual ticks that have expired so all wakeup sources reflect the current time
+            ExpireTicks();
+
+            RetransTable[i].rc                  = rc;
+            RetransTable[i].msgId               = messageId;
+            RetransTable[i].msgBuf              = msgBuf;
+            RetransTable[i].msgSendFlags        = msgSendFlags;
+            RetransTable[i].sendCount           = 0;
+            RetransTable[i].nextRetransTimeTick = static_cast<uint16_t>(
```

Comment:
I assume the cast down-casts some larger value (32 or 64 bit?) to 16. Could we have a comment on why this is safe? I imagine tick counters to be potentially large values.